scheduler/src/cook/scheduler/scheduler.clj

```diff
@@ -1232,16 +1238,43 @@
+                ; The fraction of jobs we tried to match that actually matched.
+                fraction-unmatched (if (pos? number-total) (/ (float number-unmatched) number-total) 0)
+                ; We want to autoscale any unmatched job, or
+                ; We want to scale our max-jobs-for-autoscaling by the fraction of the jobs we just matched.
+                ; E.g. If we matched 20% of the queue, then we want to autoscale to 20% of max-jobs-for-autoscaling)
+                ; We include a scale factor however, so that if we match 20$ and scale factor is 2.5, we'll generate
+                ; pods for 50% of max-jobs-for-autoscaling.
+                number-for-autoscaling (-> fraction-unmatched
+                                           (* autoscaling-scale-factor)
+                                           (min 1) ; Can't match more than 100% of max-jobs-for-autoscaling.
+                                           (* max-jobs-for-autoscaling)
+                                           int
+                                           (max number-unmatched)) ; Autoscale at least the pods that failed to match.
                 ;; We need to filter pending jobs based on quota so that we don't
                 ;; trigger autoscaling beyond what users have quota to actually run
                 autoscalable-jobs (->> pool-name
                                        (get @pool-name->pending-jobs-atom)
                                        (tools/filter-pending-jobs-for-quota pool-name (atom {}) (atom {})
-                                         user->quota user->usage (tools/global-pool-quota (config/pool-quotas) pool-name)))]
+                                                                            user->quota user->usage (tools/global-pool-quota (config/pool-quotas) pool-name))
+                                       (take number-for-autoscaling))
+                filtered-autoscalable-jobs (remove #(.getIfPresent caches/autoscale-retry-blacklist (:job/uuid %)) autoscalable-jobs)]
+            ; When we have at least 20 jobs being looked at, metric which fraction have matched.
+            (when (> number-total 20)
+              (histograms/update! (histograms/histogram (metric-title "fraction-unmatched" pool-name)) fraction-unmatched))
+            (when (pos? number-total)
+              (log/info "In" pool-name "autoscaling variables" {:number-total number-total
+                                                                :fraction-unmatched fraction-unmatched
+                                                                :number-for-autoscaling number-for-autoscaling
+                                                                :autoscalable-jobs (count autoscalable-jobs)
+                                                                :filtered-autoscalable-jobs (count filtered-autoscalable-jobs)
+                                                                :number-unmatched number-unmatched}))
```

Comment:
Should we sort these lexicographically in the map?

Reply:
Ok.